ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java

```diff
@@ -290,7 +291,8 @@ public void onMatch(RelOptRuleCall call) {
       RelNode r = relBuilder.build();
       RelOptCost afterCost = mq.getCumulativeCost(r);
       RelOptCost beforeCost = mq.getCumulativeCost(aggregate);
-      if (afterCost.isLt(beforeCost)) {
+      boolean shouldForceTransform = isGroupingUnique(join, aggregate.getGroupSet());
+      if (shouldForceTransform || afterCost.isLt(beforeCost)) {
```

Comment:
Additionally, when we force triggering the transform, does it make sense to verify that we are not creating an aggregate on top (i.e., we end up with agg before join and after join?
That may narrow it down even further to a case when the pushdown should always be beneficial.